File: src/lib/v4l2_device.cpp
#include "v4l2_device.h"
#include <iostream>
#include <cstring>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>

V4L2Device::V4L2Device() = default;

V4L2Device::~V4L2Device() {
    close();
}

bool V4L2Device::open(std::string_view device_path) {
    if (is_open()) {
        std::cerr << "Устройство уже открыто" << std::endl;
        return false;
    }

    fd_ = ::open(device_path.data(), O_RDWR | O_NONBLOCK, 0);
    if (fd_ < 0) {
        std::cerr << "Ошибка открытия устройства " << device_path << ": " << strerror(errno) << std::endl;
        return false;
    }

    poll_fds_.resize(1);
    poll_fds_[0].fd = fd_;
    poll_fds_[0].events = POLLPRI; // По умолчанию для событий

    std::cout << "Устройство " << device_path << " открыто, fd=" << fd_ << std::endl;
    return true;
}

void V4L2Device::close() {
    if (is_open()) {
        if (::close(fd_) < 0) {
            std::cerr << "Ошибка закрытия устройства: " << strerror(errno) << std::endl;
        }
        fd_ = -1;
        poll_fds_.clear();
        std::cout << "Устройство закрыто" << std::endl;
    }
}

bool V4L2Device::ioctl_helper(unsigned long request, void* arg, const char* request_name) {
    if (!is_open()) {
        std::cerr << "Устройство не открыто для ioctl " << request_name << std::endl;
        return false;
    }
    if (ioctl(fd_, request, arg) < 0) {
        std::cerr << "Ошибка ioctl " << request_name << ": " << strerror(errno) << std::endl;
        return false;
    }
    return true;
}

bool V4L2Device::query_capability(v4l2_capability& cap) {
    return ioctl_helper(VIDIOC_QUERYCAP, &cap, "VIDIOC_QUERYCAP");
}

bool V4L2Device::set_format(v4l2_format& fmt) {
    return ioctl_helper(VIDIOC_S_FMT, &fmt, "VIDIOC_S_FMT");
}

bool V4L2Device::get_format(v4l2_format& fmt) {
    return ioctl_helper(VIDIOC_G_FMT, &fmt, "VIDIOC_G_FMT");
}

bool V4L2Device::set_control(const v4l2_control& ctrl) {
    // Создаем копию, так как ioctl может изменять структуру
    v4l2_control temp_ctrl = ctrl;
    return ioctl_helper(VIDIOC_S_CTRL, &temp_ctrl, "VIDIOC_S_CTRL");
}

bool V4L2Device::request_buffers(v4l2_requestbuffers& req) {
    return ioctl_helper(VIDIOC_REQBUFS, &req, "VIDIOC_REQBUFS");
}

bool V4L2Device::queue_buffer(v4l2_buffer& buf) {
    return ioctl_helper(VIDIOC_QBUF, &buf, "VIDIOC_QBUF");
}

bool V4L2Device::dequeue_buffer(v4l2_buffer& buf) {
    // Этот ioctl может возвращать EAGAIN, что не является фатальной ошибкой
    if (!is_open()) {
        std::cerr << "Устройство не открыто для ioctl VIDIOC_DQBUF" << std::endl;
        return false;
    }
    if (ioctl(fd_, VIDIOC_DQBUF, &buf) < 0) {
        if (errno != EAGAIN) {
            std::cerr << "Ошибка ioctl VIDIOC_DQBUF: " << strerror(errno) << std::endl;
        }
        return false;
    }
    return true;
}

bool V4L2Device::stream_on(enum v4l2_buf_type type) {
    return ioctl_helper(VIDIOC_STREAMON, &type, "VIDIOC_STREAMON");
}

bool V4L2Device::stream_off(enum v4l2_buf_type type) {
    return ioctl_helper(VIDIOC_STREAMOFF, &type, "VIDIOC_STREAMOFF");
}

bool V4L2Device::subscribe_to_events() {
    v4l2_event_subscription sub_eos = {};
    sub_eos.type = V4L2_EVENT_EOS;
    if (!subscribe_event(sub_eos)) return false;

    v4l2_event_subscription sub_source_change = {};
    sub_source_change.type = V4L2_EVENT_SOURCE_CHANGE;
    if (!subscribe_event(sub_source_change)) return false;

    std::cout << "✅ Подписка на события V4L2_EVENT_EOS и V4L2_EVENT_SOURCE_CHANGE" << std::endl;
    return true;
}

bool V4L2Device::dequeue_event(v4l2_event& ev) {
    return ioctl_helper(VIDIOC_DQEVENT, &ev, "VIDIOC_DQEVENT");
}

bool V4L2Device::poll(short events, int timeout_ms) {
    if (!is_open()) {
        std::cerr << "Устройство не открыто для poll" << std::endl;
        return false;
    }
    poll_fds_[0].events = events;
    int ret = ::poll(poll_fds_.data(), poll_fds_.size(), timeout_ms);
    if (ret < 0) {
        std::cerr << "Ошибка poll: " << strerror(errno) << std::endl;
        revents_ = 0;
        return false;
    }
    if (ret == 0) {
        // Timeout
        revents_ = 0;
        return true;
    }
    revents_ = poll_fds_[0].revents;
    return true;
}

bool V4L2Device::has_event() const {
    return revents_ & POLLPRI;
}

bool V4L2Device::has_error() const {
    return revents_ & POLLERR;
}

bool V4L2Device::is_ready_for_read() const {
    return revents_ & POLLIN;
}

bool V4L2Device::is_ready_for_write() const {
    return revents_ & POLLOUT;
}

bool V4L2Device::subscribe_event(v4l2_event_subscription& sub) {
    return ioctl_helper(VIDIOC_SUBSCRIBE_EVENT, &sub, "VIDIOC_SUBSCRIBE_EVENT");
}

bool V4L2Device::configure_decoder_formats(uint32_t width, uint32_t height, uint32_t in_pixel_format, uint32_t out_pixel_format) {
    struct v4l2_format fmt_in = {};
    fmt_in.type = V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE;
    fmt_in.fmt.pix_mp.width = width;
    fmt_in.fmt.pix_mp.height = height;
    fmt_in.fmt.pix_mp.pixelformat = in_pixel_format;
    fmt_in.fmt.pix_mp.num_planes = 1;
    fmt_in.fmt.pix_mp.plane_fmt[0].sizeimage = 2 * 1024 * 1024; // 2MB for H.264
    if (!set_format(fmt_in)) {
        std::cerr << "❌ ОШИБКА: Не удалось установить входной формат" << std::endl;
        return false;
    }
    std::cout << "Входной формат установлен: " << width << "x" << height << std::endl;

    struct v4l2_format fmt_out = {};
    fmt_out.type = V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE;
    fmt_out.fmt.pix_mp.width = width;
    fmt_out.fmt.pix_mp.height = height;
    fmt_out.fmt.pix_mp.pixelformat = out_pixel_format;
    fmt_out.fmt.pix_mp.num_planes = 1;
    if (!set_format(fmt_out)) {
        std::cerr << "❌ ОШИБКА: Не удалось установить выходной формат" << std::endl;
        return false;
    }
    std::cout << "Выходной формат установлен: " << width << "x" << height << std::endl;

    // Устанавливаем минимальное количество буферов для захвата для минимальной задержки
    struct v4l2_control ctrl = {};
    ctrl.id = V4L2_CID_MIN_BUFFERS_FOR_CAPTURE;
    ctrl.value = 1;
    if (!set_control(ctrl)) {
        std::cout << "⚠️ ПРЕДУПРЕЖДЕНИЕ: Не удалось установить V4L2_CID_MIN_BUFFERS_FOR_CAPTURE=1. Это может увеличить задержку." << std::endl;
        // Не критично, продолжаем
    } else {
        std::cout << "✅ Установлена минимальная буферизация на выходе (MIN_BUFFERS_FOR_CAPTURE=1) для низкой задержки" << std::endl;
    }

    return true;
}

bool V4L2Device::check_dma_buf_support() {
    // Пытаемся запросить буферы в режиме DMA-buf для тестирования
    struct v4l2_requestbuffers req = {};
    req.count = 1;
    req.type = V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE;
    req.memory = V4L2_MEMORY_DMABUF;

    bool supported = request_buffers(req);

    std::cout << "Проверка поддержки DMA-buf: " << (supported ? "OK" : "FAIL")
              << std::endl;

    // Сбрасываем буферы после тестирования
    if (supported) {
        req.count = 0;
        (void)request_buffers(req);
    }

    return supported;
}

bool V4L2Device::initialize_for_decoding(std::string_view device_path) {
    std::cout << "Инициализация V4L2 устройства для декодирования: " << device_path << std::endl;

    if (!open(device_path)) {
        return false;
    }

    struct v4l2_capability cap;
    if (!query_capability(cap)) {
        close();
        return false;
    }

    std::cout << "Устройство: " << cap.card << "\nДрайвер: " << cap.driver << std::endl;

    if (!(cap.capabilities & V4L2_CAP_VIDEO_M2M_MPLANE)) [[unlikely]] {
        std::cerr << "❌ ОШИБКА: Устройство не поддерживает V4L2_CAP_VIDEO_M2M_MPLANE" << std::endl;
        std::cerr << "Доступные возможности: 0x" << std::hex << cap.capabilities << std::dec << std::endl;
        close();
        return false;
    }

    // Проверяем поддержку DMA-buf в V4L2 драйвере
    if (!check_dma_buf_support()) {
        std::cerr << "❌ КРИТИЧЕСКАЯ ОШИБКА: V4L2 драйвер не поддерживает DMA-buf" << std::endl;
        std::cerr << "Возможные причины:" << std::endl;
        std::cerr << "  - Устаревший драйвер" << std::endl;
        std::cerr << "  - Неправильная конфигурация ядра" << std::endl;
        close();
        return false;
    }

    std::cout << "Используем DMA-buf буферы" << std::endl;

    // Подписываемся на критические V4L2 события
    if (!subscribe_to_events()) {
        std::cerr << "⚠️ ПРЕДУПРЕЖДЕНИЕ: Не удалось подписаться на V4L2 события" << std::endl;
        // Продолжаем работу - события не критичны для базовой функциональности
    }

    return true;
}

File: cs-src/RtpDrmPlayer/Core/V4L2Device.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using RtpDrmPlayer.Native;

namespace RtpDrmPlayer.Core;

/// <summary>
/// Класс V4L2Device инкапсулирует низкоуровневое взаимодействие с устройством V4L2.
/// Обеспечивает тонкую обертку вокруг вызовов ioctl для V4L2, управляя файловым дескриптором 
/// и выполняя базовые операции с устройством. Полностью соответствует реализации C++.
/// </summary>
public sealed class V4L2Device : IDisposable
{
    private int _fd = -1;
    private PollFd[] _pollFds = Array.Empty<PollFd>();
    private short _revents;

    public V4L2Device()
    {
        Console.WriteLine("V4L2Device created");
    }

    ~V4L2Device() => Close();

    /// <summary>
    /// Открывает устройство V4L2
    /// </summary>
    /// <param name="devicePath">Путь к устройству</param>
    /// <returns>true при успешном открытии</returns>
    public bool Open(string devicePath)
    {
    if (IsOpen)
        {
            Console.WriteLine("Device already open");
            return false;
        }
        
        Console.WriteLine($"Opening V4L2 device: {devicePath}");
        
        _fd = LibC.open(devicePath, LibC.O_RDWR | LibC.O_NONBLOCK, 0);
        
        if (_fd < 0)
        {
            Console.WriteLine($"❌ Failed to open {devicePath}, errno: {Marshal.GetLastWin32Error()}");
            return false;
        }
        
        _pollFds = new[]
        {
            new PollFd
            {
                Fd = _fd,
                Events = LibC.POLLPRI  // Инициализируем для событий по умолчанию
            }
        };
        
        Console.WriteLine($"✅ Opened V4L2 device with fd {_fd}");
        return true;
    }

    /// <summary>
    /// Закрывает устройство V4L2
    /// </summary>
    public void Close()
    {
    if (!IsOpen)
            return;
            
        Console.WriteLine($"Closing V4L2 device with fd {_fd}");
        
        LibC.close(_fd);
        _fd = -1;
        _pollFds = Array.Empty<PollFd>();
        
        Console.WriteLine("V4L2 device closed");
    }

    /// <summary>
    /// Проверяет, открыто ли устройство
    /// </summary>
    public bool IsOpen => _fd >= 0;

    /// <summary>
    /// Возвращает файловый дескриптор устройства
    /// </summary>
    public int Fd() => _fd;

    /// <summary>
    /// Вспомогательный метод для выполнения ioctl с подробной обработкой ошибок
    /// </summary>
    private bool ioctl_helper(ulong request, IntPtr arg, string requestName)
    {
    if (!IsOpen)
        {
            Console.WriteLine($"❌ Device not open for {requestName}");
            return false;
        }
        
        int result = LibC.ioctl(_fd, request, arg);
        
        if (result != 0)
        {
            int errno = Marshal.GetLastWin32Error();
            Console.WriteLine($"❌ {requestName} failed: errno={errno}");
            return false;
        }
        
        return true;
    }
    
    // Безопасная версия ioctl с использованием GCHandle
    private unsafe bool SafeIoctl<T>(ulong request, ref T structure, string requestName) where T : unmanaged
    {
        fixed (T* p = &structure)
        {
            return ioctl_helper(request, (IntPtr)p, requestName);
        }
    }
    
    // Оригинальная реализация для обратной совместимости
    // Removed legacy Ioctl wrapper (unused)

    /// <summary>
    /// Запрашивает информацию о возможностях устройства
    /// </summary>
    public unsafe bool QueryCapability(ref v4l2_capability cap)
    {
        bool result = SafeIoctl(V4L2Const.VIDIOC_QUERYCAP, ref cap, "VIDIOC_QUERYCAP");
        
        if (result)
        {
            string driverName, cardName, busInfo;
            
            fixed (byte* pDriver = cap.driver)
            {
                driverName = Marshal.PtrToStringAnsi((IntPtr)pDriver) ?? string.Empty;
            }
            
            fixed (byte* pCard = cap.card)
            {
                cardName = Marshal.PtrToStringAnsi((IntPtr)pCard) ?? string.Empty;
            }

            fixed (byte* pBusInfo = cap.bus_info)
            {
                busInfo = Marshal.PtrToStringAnsi((IntPtr)pBusInfo) ?? string.Empty;
            }
            
            Console.WriteLine($"Device capabilities: driver='{driverName}', card='{cardName}', bus='{busInfo}'");
        }
        
        return result;
    }

    /// <summary>
    /// Устанавливает формат для видео потока
    /// </summary>
    public bool SetFormat(ref v4l2_format format)
    {
        string typeName = format.type switch
        {
            V4L2Const.V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE => "OUTPUT",
            V4L2Const.V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE => "CAPTURE",
            _ => $"Unknown({format.type})"
        };
        
        Console.WriteLine($"Setting format for {typeName}");
        
        bool result = SafeIoctl(V4L2Const.VIDIOC_S_FMT, ref format, "VIDIOC_S_FMT");
        
        if (result && format.type == V4L2Const.V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE)
        {
            Console.WriteLine($"✅ Format set: {format.fmt.width}x{format.fmt.height}, " + 
                             $"format=0x{format.fmt.pixelformat:X}, planes={format.fmt.num_planes}");
        }
        
        return result;
    }

    /// <summary>
    /// Получает текущий формат видео потока
    /// </summary>
    public bool GetFormat(ref v4l2_format format)
    {
        bool result = SafeIoctl(V4L2Const.VIDIOC_G_FMT, ref format, "VIDIOC_G_FMT");
        
        if (result)
        {
            Console.WriteLine($"Format info: {format.fmt.width}x{format.fmt.height}, " + 
                             $"format=0x{format.fmt.pixelformat:X}, planes={format.fmt.num_planes}");
        }
        
        return result;
    }

    /// <summary>
    /// Устанавливает контрольный параметр устройства
    /// </summary>
    public unsafe bool SetControl(uint id, int value)
    {
        var control = new v4l2_control { id = id, value = value };
        
        Console.WriteLine($"Setting control: id=0x{id:X}, value={value}");
        bool result = ioctl_helper(V4L2Const.VIDIOC_S_CTRL, (IntPtr)(&control), "VIDIOC_S_CTRL");
        
        if (!result)
        {
            Console.WriteLine($"❌ Failed to set control 0x{id:X}");
        }
        
        return result;
    }

    /// <summary>
    /// Запрашивает буферы указанного типа и памяти
    /// </summary>
    public bool RequestBuffers(ref v4l2_requestbuffers request)
    {
        string memType = request.memory switch
        {
            V4L2Const.V4L2_MEMORY_MMAP => "MMAP",
            V4L2Const.V4L2_MEMORY_USERPTR => "USERPTR",
            V4L2Const.V4L2_MEMORY_DMABUF => "DMABUF",
            _ => $"Unknown({request.memory})"
        };
        
        Console.WriteLine($"Requesting {request.count} buffers of type {request.type}, memory={memType}");
        bool result = SafeIoctl(V4L2Const.VIDIOC_REQBUFS, ref request, "VIDIOC_REQBUFS");
        
        if (result)
        {
            Console.WriteLine($"✅ Successfully requested {request.count} buffers");
        }
        
        return result;
    }

    /// <summary>
    /// Ставит буфер в очередь для обработки
    /// </summary>
    public bool QueueBuffer(ref v4l2_buffer buffer)
    {
        Console.WriteLine($"Queuing buffer index={buffer.index}, type={buffer.type}");
        bool result = SafeIoctl(V4L2Const.VIDIOC_QBUF, ref buffer, "VIDIOC_QBUF");
        
        if (!result)
        {
            Console.WriteLine($"❌ Failed to queue buffer {buffer.index}");
        }
        
        return result;
    }

    /// <summary>
    /// Извлекает буфер из очереди обработанных
    /// </summary>
    public bool DequeueBuffer(ref v4l2_buffer buffer)
    {
        bool result = SafeIoctl(V4L2Const.VIDIOC_DQBUF, ref buffer, "VIDIOC_DQBUF");
        
        if (result)
        {
            Console.WriteLine($"Dequeued buffer index={buffer.index}, flags=0x{buffer.flags:X}");
        }
        
        return result;
    }

    /// <summary>
    /// Включает поток указанного типа
    /// </summary>
    public unsafe bool StreamOn(uint type)
    {
        uint typeVal = type;
        
        Console.WriteLine($"Starting stream for type {type}");
        bool result = ioctl_helper(V4L2Const.VIDIOC_STREAMON, (IntPtr)(&typeVal), "VIDIOC_STREAMON");
        
        if (!result)
        {
            Console.WriteLine($"❌ Failed to start stream for type {type}");
        }
        else
        {
            Console.WriteLine($"✅ Stream started for type {type}");
        }
        
        return result;
    }

    /// <summary>
    /// Выключает поток указанного типа
    /// </summary>
    public unsafe bool StreamOff(uint type)
    {
        uint typeVal = type;
        
        Console.WriteLine($"Stopping stream for type {type}");
        bool result = ioctl_helper(V4L2Const.VIDIOC_STREAMOFF, (IntPtr)(&typeVal), "VIDIOC_STREAMOFF");
        
        if (!result)
        {
            Console.WriteLine($"❌ Failed to stop stream for type {type}");
        }
        else
        {
            Console.WriteLine($"✅ Stream stopped for type {type}");
        }
        
        return result;
    }

    /// <summary>
    /// Подписка на определенное событие устройства
    /// </summary>
    public bool SubscribeEvent(ref v4l2_event_subscription subscription)
    {
        Console.WriteLine($"Subscribing to event type {subscription.type}");
        bool result = SafeIoctl(V4L2Const.VIDIOC_SUBSCRIBE_EVENT, ref subscription, "VIDIOC_SUBSCRIBE_EVENT");
        
        if (!result)
        {
            Console.WriteLine($"❌ Failed to subscribe to event {subscription.type}");
        }
        
        return result;
    }

    /// <summary>
    /// Подписывается на все типы событий декодера
    /// </summary>
    public bool SubscribeToEvents()
    {
        // Подписываемся на все события кодека/декодера
        var sub_src_change = new v4l2_event_subscription
        {
            type = V4L2Const.V4L2_EVENT_SOURCE_CHANGE
        };
        
    if (!SubscribeEvent(ref sub_src_change))
        {
            Console.WriteLine("❌ Failed to subscribe to SOURCE_CHANGE events");
            return false;
        }

        var sub_eos = new v4l2_event_subscription
        {
            type = V4L2Const.V4L2_EVENT_EOS
        };
    if (!SubscribeEvent(ref sub_eos))
        {
            Console.WriteLine("❌ Failed to subscribe to EOS events");
            return false;
        }
        
        Console.WriteLine("✅ Successfully subscribed to decoder events");
        return true;
    }

    /// <summary>
    /// Получает событие из очереди событий
    /// </summary>
    public bool DequeueEvent(ref v4l2_event ev)
    {
        bool result = SafeIoctl(V4L2Const.VIDIOC_DQEVENT, ref ev, "VIDIOC_DQEVENT");
        
        if (result)
        {
            Console.WriteLine($"Dequeued event type={ev.type}");
        }
        
        return result;
    }

    /// <summary>
    /// Проверка ожидающих событий или готовности операций ввода-вывода
    /// </summary>
    public bool Poll(short events, int timeoutMs)
    {
    if (!IsOpen)
        {
            Console.WriteLine("❌ Cannot poll: device not open");
            return false;
        }
        
        _pollFds[0].Events = events;
        
        int result = LibC.poll(_pollFds, (uint)_pollFds.Length, timeoutMs);
        
        if (result < 0)
        {
            int errno = Marshal.GetLastWin32Error();
            Console.WriteLine($"❌ Poll failed: errno={errno}");
            _revents = 0;
            return false;
        }
        
        _revents = _pollFds[0].Revents;
        
        if (_revents != 0)
        {
            Console.WriteLine($"Poll returned events: 0x{_revents:X}");
        }
        
        return true;
    }

    /// <summary>
    /// Проверяет наличие события на устройстве
    /// </summary>
    public bool HasEvent => (_revents & LibC.POLLPRI) != 0;

    /// <summary>
    /// Проверяет наличие ошибки на устройстве
    /// </summary>
    public bool HasError => (_revents & LibC.POLLERR) != 0;

    /// <summary>
    /// Проверяет готовность устройства к чтению
    /// </summary>
    public bool ReadyRead => (_revents & LibC.POLLIN) != 0;

    /// <summary>
    /// Проверяет готовность устройства к записи
    /// </summary>
    public bool ReadyWrite => (_revents & LibC.POLLOUT) != 0;

    /// <summary>
    /// Настраивает форматы для декодирования
    /// </summary>
    public bool ConfigureDecoderFormats(uint width, uint height, uint inPixelFormat, uint outPixelFormat)
    {
    return ConfigureDecoderFormats(width, height, inPixelFormat, outPixelFormat, width * height); // По умолчанию размер входного буфера
    }

    /// <summary>
    /// Настраивает форматы для декодирования с указанным размером входного буфера
    /// </summary>
    public bool ConfigureDecoderFormats(uint width, uint height, uint inPixelFormat, uint outPixelFormat, uint inputSizeImage)
    {
        Console.WriteLine($"Configuring decoder formats: {width}x{height}, in=0x{inPixelFormat:X}, out=0x{outPixelFormat:X}");
        
        // Настраиваем входной формат (OUTPUT)
        var inputFormat = new v4l2_format
        {
            type = V4L2Const.V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE
        };
        inputFormat.fmt.width = width;
        inputFormat.fmt.height = height;
        inputFormat.fmt.pixelformat = inPixelFormat;
        inputFormat.fmt.num_planes = 1;
        inputFormat.fmt.plane_fmt_0.sizeimage = inputSizeImage;
        inputFormat.fmt.plane_fmt_0.bytesperline = 0;
        
    if (!SetFormat(ref inputFormat))
        {
            Console.WriteLine("❌ Failed to set input format");
            return false;
        }
        
        // Настраиваем выходной формат (CAPTURE)
        var outputFormat = new v4l2_format
        {
            type = V4L2Const.V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE
        };
        outputFormat.fmt.width = width;
        outputFormat.fmt.height = height;
        outputFormat.fmt.pixelformat = outPixelFormat;
        outputFormat.fmt.num_planes = 1; // Для NV12 будет 2, но драйвер может сам исправить
        // Для NV12, sizeimage это общий размер, а bytesperline - ширина luma
        outputFormat.fmt.plane_fmt_0.sizeimage = width * height * 3 / 2; 
        outputFormat.fmt.plane_fmt_0.bytesperline = (ushort)width;
        
    if (!SetFormat(ref outputFormat))
        {
            Console.WriteLine("❌ Failed to set output format");
            return false;
        }
        
        Console.WriteLine("✅ Successfully configured decoder formats");
        return true;
    }

    /// <summary>
    /// Проверяет поддержку DMA-BUF
    /// </summary>
    public bool CheckDmaBufSupport()
    {
        Console.WriteLine("Checking DMA-BUF support...");
        
        var request = new v4l2_requestbuffers
        {
            count = 1,
            type = V4L2Const.V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE,
            memory = V4L2Const.V4L2_MEMORY_DMABUF
        };
        
    if (!RequestBuffers(ref request))
        {
            Console.WriteLine("❌ DMA-BUF not supported");
            return false;
        }
        
        // Освобождаем запрошенные буферы
        request.count = 0;
    RequestBuffers(ref request);
        
        Console.WriteLine("✅ DMA-BUF is supported");
        return true;
    }
    
    /// <summary>
    /// Инициализирует устройство для декодирования
    /// </summary>
    public bool InitializeForDecoding(string devicePath)
    {
        Console.WriteLine($"Initializing decoder on {devicePath}");
        
    if (!Open(devicePath))
        {
            Console.WriteLine("❌ Failed to open device");
            return false;
        }
        
        // Проверка возможностей
        var capability = new v4l2_capability();
    if (!QueryCapability(ref capability))
        {
            Console.WriteLine("❌ Failed to query capabilities");
            Close();
            return false;
        }
        
        // Проверка поддержки DMA-BUF
    if (!CheckDmaBufSupport())
        {
            Console.WriteLine("❌ DMA-BUF not supported");
            Close();
            return false;
        }
        
        // Подписка на события
    if (!SubscribeToEvents())
        {
            Console.WriteLine("❌ Failed to subscribe to events");
            Close();
            return false;
        }
        
        Console.WriteLine("✅ Device initialized for decoding");
        return true;
    }

    /// <summary>
    /// Ставит многоплоскостной DMA-BUF буфер в очередь
    /// </summary>
    public bool QueueMultiPlaneDmabuf(uint index, uint type, int[] fds, uint[] lengths, uint[] dataOffsets) =>
        QueueMultiPlaneDmabuf(index, type, fds, lengths, null, dataOffsets, 0);

    /// <summary>
    /// Ставит многоплоскостной DMA-BUF буфер в очередь с дополнительными параметрами
    /// </summary>
    public unsafe bool QueueMultiPlaneDmabuf(uint index, uint type, int[] fds, uint[] lengths, uint[]? bytesused, uint[] dataOffsets, uint flags = 0)
    {
        if (fds == null || lengths == null || dataOffsets == null)
        {
            Console.WriteLine("❌ Invalid parameters: arrays cannot be null");
            return false;
        }
        
        int planesCount = fds.Length;
        
        if (planesCount == 0 || lengths.Length != planesCount || dataOffsets.Length != planesCount)
        {
            Console.WriteLine("❌ Invalid parameters: array lengths must match");
            return false;
        }
        
        if (bytesused != null && bytesused.Length != planesCount)
        {
            Console.WriteLine("❌ Invalid parameters: bytesused array length must match planes");
            return false;
        }
        
        Console.WriteLine($"Queueing DMA-BUF buffer: index={index}, type={type}, planes={planesCount}, flags=0x{flags:X}");
        
        var planes = new v4l2_plane[planesCount];
        for (int i = 0; i < planesCount; i++)
        {
            planes[i] = new v4l2_plane
            {
                bytesused = bytesused?[i] ?? 0,
                length = lengths[i],
                data_offset = dataOffsets[i]
            };
            planes[i].m_fd = fds[i];
            Console.WriteLine($"  Plane {i}: fd={fds[i]}, length={lengths[i]}, bytesused={planes[i].bytesused}, offset={dataOffsets[i]}");
        }

        fixed (v4l2_plane* pPlanes = planes)
        {
            var buffer = new v4l2_buffer
            {
                index = index,
                type = type,
                memory = V4L2Const.V4L2_MEMORY_DMABUF,
                length = (uint)planesCount,
                flags = flags,
                m_planes = (IntPtr)pPlanes
            };
            return QueueBuffer(ref buffer);
        }
    }

    /// <summary>
    /// Ставит одноплоскостной DMA-BUF буфер в очередь
    /// </summary>
    public unsafe bool QueueSinglePlaneDmabuf(uint index, uint type, int fd, uint bytesused, uint length, uint flags = 0)
    {
        Console.WriteLine($"Queueing single-plane DMA-BUF buffer: index={index}, type={type}, fd={fd}, bytesused={bytesused}, flags=0x{flags:X}");

        var plane = new v4l2_plane
        {
            bytesused = bytesused,
            length = length,
            data_offset = 0
        };
        plane.m_fd = fd;

        var buffer = new v4l2_buffer
        {
            index = index,
            type = type,
            memory = V4L2Const.V4L2_MEMORY_DMABUF,
            length = 1, // Single plane
            flags = flags,
            m_planes = (IntPtr)(&plane)
        };

    return QueueBuffer(ref buffer);
    }

    /// <summary>
    /// Извлекает многоплоскостной буфер из очереди обработанных
    /// </summary>
    public unsafe bool DequeueMultiPlane(uint type, out uint index, out uint flags, out uint sequence, out v4l2_plane[] planes)
    {
        index = flags = sequence = 0;
        planes = Array.Empty<v4l2_plane>();
        
        // Предполагаем максимум 4 плоскости (обычно для NV12/YUV420 и подобных)
        int maxPlanes = 4;
        var planeArray = new v4l2_plane[maxPlanes];

        fixed (v4l2_plane* pPlanes = planeArray)
        {
            var buffer = new v4l2_buffer
            {
                type = type,
                memory = V4L2Const.V4L2_MEMORY_DMABUF,
                length = (uint)maxPlanes,
                m_planes = (IntPtr)pPlanes
            };
            
            if (!DequeueBuffer(ref buffer))
            {
                // Console.WriteLine("❌ Failed to dequeue buffer"); // This is too noisy for non-blocking calls
                return false;
            }
            
            index = buffer.index;
            flags = buffer.flags;
            sequence = buffer.sequence;
            
            int actualPlanes = (int)buffer.length;
            if (actualPlanes > maxPlanes)
            {
                actualPlanes = maxPlanes;
            }
            
            planes = new v4l2_plane[actualPlanes];
            Array.Copy(planeArray, planes, actualPlanes);
            
            // Console.WriteLine($"Dequeued buffer {index} with {actualPlanes} planes, sequence={sequence}");
            return true;
        }
    }
    
    /// <summary>
    /// Методы совместимости с предыдущей C# версией
    /// </summary>
    // Removed legacy compatibility wrappers after renaming core methods to PascalCase.

    /// <summary>
    /// Освобождает ресурсы
    /// </summary>
    public void Dispose() => Close();
}
